app/models/user.rb

```diff
@@ -2,47 +2,82 @@
 
 class User < ApplicationRecord
   # Include default devise modules. Others available are:
-  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  # :confirmable, :lockable, :timeoutable, :trackable and
   default_scope { order(created_at: :desc) }
   before_save :capitalize_names
-  validates :lastname, presence: true
-  validates :firstname, presence: true
-  validates :dob, presence: true
-  validates :gender, presence: true
 
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: [:facebook]
 
   has_many :posts, dependent: :destroy
   has_many_attached :avatar, dependent: :destroy
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
   has_many :liked_posts, through: :likes, source: 'post'
   has_many :friendships
+  has_many :received_but_unconfirmed, -> { where confirmed: false }, class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :pending_friends, through: :received_but_unconfirmed, source: 'user'
   has_many :inverted_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
   scope :all_except, ->(me) { where.not(id: me) }
+  has_many :confirmed_friendships, -> { where confirmed: true }, class_name: 'Friendship'
+  has_many :friends, through: :confirmed_friendships, source: 'user'
+
   def names
     "#{lastname} #{firstname}"
   end
 
-  def all_friends
+  def sent_and_received_requests
     receive_request = inverted_friendships.where(confirmed: 'true')
     sent_request = friendships.where(confirmed: 'true')
     receive_request + sent_request
   end
 
   def approved_friend
-    all_friends.map(&:user_id)
+    me = sent_and_received_requests.map(&:user_id)
+    you = sent_and_received_requests.map(&:friend_id)
+    (me + you).uniq - [id]
+  end
+
+  def mutual_friend(myfriend)
+    me = approved_friend
+    you = myfriend.approved_friend
+    me & you
+  end
+
+  def mutual_friend_count(myfriend)
+    mutual_friend(myfriend).count
   end
 
   def find_friend(hello)
     User.find(hello)
   end
 
+  def and_friends
+    Post.where(user_id: self) + Post.where(user_id: friends)
+  end
+
+  def pending_friends_arr
+    pending_friends.ids
+  end
+
+  def approved_friends_arr
+    friends.ids
+  end
+
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+      user.email = auth.info.email
+      user.password = Devise.friendly_token[0, 20]
+      user.lastname = auth.info.last_name
+      user.firstname = auth.info.first_name
```

Comment:
[OPTIONAL]
Something like this could also work
```ruby
      user.first_name = auth.info.name.split.first
      user.last_name = auth.info.name.split.last
```

Reply:
#### Fixed issue with something wrong user can now sign in with facebook account smoothly